Add `DiscoveryURL` to Authentication Configuration

aramase opened the pull request:
/kind feature
/kind api-change

- Add `DiscoveryURL` to Authentication Configuration.

```release-note
Structured Authentication Configuration now supports `DiscoveryURL`. 
discoveryURL if specified, overrides the URL used to fetch discovery information. 
This is for scenarios where the well-known and jwks endpoints are hosted at a different
location than the issuer (such as locally in the cluster).
```

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3331-structured-authentication-configuration
```

aramase committed:
Add `DiscoveryURL` to AuthenticationConfiguration

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase committed:
wire up discovery url in authenticator

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase commented:
/sig auth
/triage accepted
/priority important-soon

aramase commented:
/assign enj liggitt

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -175,9 +175,10 @@ type JWTAuthenticator struct {
 }
 
-// Issuer provides the configuration for a external provider specific settings.
+// Issuer provides the configuration for an external provider specific settings.
```
```suggestion
// Issuer provides the configuration for an external provider's specific settings.
```

Also update this go doc in `staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go`.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +351,27 @@ func New(opts Options) (*Authenticator, error) {
+	discoveryURL *url.URL
+}
+
+func (t *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if t.discoveryURL != nil {
+		if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, "/.well-known/openid-configuration") {
+			req.URL.Host = t.discoveryURL.Host
+			req.URL.Scheme = t.discoveryURL.Scheme
+		}
+	}
+	return t.rt.RoundTrip(req)
+}
```
This isn't correct for cases where the discovery and/or the issuer URL have a path component that is different.

i.e.

issuer URL = `https://hello.com/a/b/foo`
discovery URL = `https://bye.com/c/d/bar`

You will see a request for

`https://hello.com/a/b/foo/.well-known/openid-configuration`

and rewrite it to

`https://bye.com/a/b/foo/.well-known/openid-configuration`

instead of the correct URL

`https://bye.com/c/d/bar/.well-known/openid-configuration`

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +351,27 @@ func New(opts Options) (*Authenticator, error) {
+	discoveryURL *url.URL
+}
+
+func (t *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```
Clone the request before mutating it.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +351,27 @@ func New(opts Options) (*Authenticator, error) {
+	if t.discoveryURL != nil {
+		if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, "/.well-known/openid-configuration") {
+			req.URL.Host = t.discoveryURL.Host
+			req.URL.Scheme = t.discoveryURL.Scheme
```
Scheme should never change, it should always be `https`.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +351,27 @@ func New(opts Options) (*Authenticator, error) {
 
+// roundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type roundTripper struct {
```
Give this struct a more specific name.

aramase replied:
sounds good! I was thinking I'll use the same one for the jwks metrics and hence the generic name, but I can just have different roundTrippers each scoped to a specific task and layered?

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -100,21 +100,33 @@ func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field
+	return validateURL(issuerURL, fldPath)
+}
+
+func validateIssuerDiscoveryURL(issuerDiscoveryURL string, fldPath *field.Path) field.ErrorList {
```
These need to be unique right?  Also probably not the same as the configured issuer URL?

aramase replied:
They need to be unique but as of this PR we still only support 1 JWT authenticator. I've a PR open to support multiple https://github.com/kubernetes/kubernetes/pull/123431 and added the issuer unique check there. I was waiting to see which PR to add the uniqueness check for discovery url based on which one got merged first.

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -100,21 +100,33 @@ func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field
 	var allErrs field.ErrorList
 
-	allErrs = append(allErrs, validateURL(issuer.URL, fldPath.Child("url"))...)
+	allErrs = append(allErrs, validateIssuerURL(issuer.URL, fldPath.Child("url"))...)
```
Where is the unique issuer check?

aramase replied:
They need to be unique but as of this PR we still only support 1 JWT authenticator. I've a PR open to support multiple https://github.com/kubernetes/kubernetes/pull/123431 and added the issuer unique check there.

enj commented on test/integration/apiserver/oidc/oidc_test.go:
```diff
@@ -840,6 +840,70 @@ jwt:
+kind: AuthenticationConfiguration
+jwt:
+- issuer:
+    url: https://example.com
+    discoveryURL: %s
```
Would be good to exercise issuers and discovery URLs with path components.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:
```diff
@@ -306,7 +311,19 @@ func (c *claimsTest) run(t *testing.T) {
-	got, ok, err := a.AuthenticateToken(testContext(t), token)
+	ctx := testContext(t)
+	// wait for the authenticator to be initialized
+	err = wait.PollUntilContextCancel(ctx, time.Millisecond, true, func(context.Context) (bool, error) {
+		if v, _ := a.idTokenVerifier(); v == nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to initialize the authenticator: %v", err)
+	}
```
Uh what changed to need this?

aramase replied:
With the `"discovery-url"` unit test, I'm trying to initialize the provider by fetching keys from the claims server (`fetchKeysFromRemote`) and skipping configuring a static key set [here](https://github.com/kubernetes/kubernetes/pull/123527/files#diff-9dfc52cf08b2714b4e54cfc514504fbf637c9182648855e7bc7a3f210d6ea63dR273-R276). With this for the test, the provider is initialized async and there is some delay. This is to ensure the provider is ready before calling `AuthenticateToken`.

I had this wrapped in a `if c.fetchKeysFromRemote` but figured this check shouldn't have any impact on the static key set cases either.

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:
```diff
@@ -2953,6 +2970,43 @@ func TestToken(t *testing.T) {
+			options: Options{
+				JWTAuthenticator: apiserver.JWTAuthenticator{
+					Issuer: apiserver.Issuer{
+						URL:          "https://auth.example.com",
+						DiscoveryURL: "{{.URL}}",
```
Need tests with URLs that have path components.

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -100,21 +100,36 @@ func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field
+	if len(issuerDiscoveryURL) == 0 {
+		return nil
+	}
+	if len(issuerURL) > 0 && issuerURL == issuerDiscoveryURL {
```
I think we should probably ignore trailing slashes when comparing these URLs?

`https://a.com` vs `https://a.com/` ?

enj replied:
Also I do not think there is any need to allow trailing slash on the discovery URL?

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +355,29 @@ func New(opts Options) (*Authenticator, error) {
+		if t.discoveryURL != nil {
+			clone := net.CloneRequest(req)
+			clone.URL.Host = t.discoveryURL.Host
+			clone.URL.Path = strings.TrimSuffix(t.discoveryURL.Path, "/") + wellKnownEndpointPath
```
Once you make it so discovery URL cannot end in a slash, this can be simplified:

```suggestion
			clone.URL.Path = t.discoveryURL.Path + wellKnownEndpointPath
```

enj commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -268,6 +273,18 @@ func New(opts Options) (*Authenticator, error) {
+		if err != nil {
+			return nil, fmt.Errorf("oidc: invalid discovery URL: %w", err)
+		}
+		client.Transport = &discoveryURLRoundTripper{client.Transport, discoveryURL}
```
Hmm since the `client` can be passed in we probably should do a shallow copy and handle the transport being `nil`.

enj commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -100,21 +100,43 @@ func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field
+	if strings.HasSuffix(issuerDiscoveryURL, "/") {
+		allErrs = append(allErrs, field.Invalid(fldPath, issuerDiscoveryURL, "discoveryURL must not contain trailing slash"))
+	}
+	if len(issuerURL) > 0 && issuerURL == strings.TrimRight(issuerDiscoveryURL, "/") {
```
Should we trim `issuerURL` also before making the check?

enj commented:
/lgtm
/approve

aramase commented:
`pull-kubernetes-unit` failure unrelated to this PR.

```
{Failed;  === RUN   TestPrintIPAddressList
    printers_test.go:6616: mismatch:   []v1.TableRow{
          	{Cells: {string("192.168.2.2"), string("myresource.mygroup/mynamespace/myname"), string("10y")}},
          	{
          		Cells: []any{
          			string("2001:db8::2"),
          			string("myresource2.mygroup2/mynamespace2/myname2"),
        - 			string("5y1d"),
        + 			string("5y2d"),
          		},
          		Conditions: nil,
          		Object:     {},
          	},
          }
--- FAIL: TestPrintIPAddressList (0.00s)
;}
```

aramase commented:
/retest

liggitt commented:
leap-day unit test failure, fix in https://github.com/kubernetes/kubernetes/pull/123603

aramase commented:
> leap-day unit test failure, fix in #123603

rebased the PR

enj commented:
/lgtm

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -220,6 +220,26 @@ type Issuer struct {
+	// discoveryURL if specified, overrides the URL used to fetch discovery information.
+	// This is for scenarios where the well-known and jwks endpoints are hosted at a different
+	// location than the issuer (such as locally in the cluster).
+	// Format must be https://url/path.
```
the issuer returned in the content when fetching `.discoveryURL` must match `.url`?

aramase replied:
done!

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go:
```diff
@@ -100,21 +100,43 @@ func validateJWTAuthenticator(authenticator api.JWTAuthenticator, fldPath *field
 	}
 
+	if strings.HasSuffix(issuerDiscoveryURL, "/") {
+		allErrs = append(allErrs, field.Invalid(fldPath, issuerDiscoveryURL, "discoveryURL must not contain trailing slash"))
```
curious where this restriction comes from

aramase replied:
This is just a restriction we're imposing so we can cleanly construct the `discoveryUrl + /.well-known/openid-configuration` post validation. We don't have any such validations for `issuer.url` because the slash could be in `issuer` in discovery doc.

liggitt commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +362,28 @@ func New(opts Options) (*Authenticator, error) {
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, wellKnownEndpointPath) {
+		clone := req.Clone(req.Context())
+		clone.URL.Host = t.discoveryURL.Host
+		clone.URL.Path = t.discoveryURL.Path + wellKnownEndpointPath
+		clone.URL.RawPath = ""
```
why not override the whole request URL to discoveryURL, and require discoveryURL to include the well-known path suffix?

as-is, you couldn't use discoveryURL to point to a location for the issuer info that didn't end in a `/.well-known/openid-configuration` suffix. That seems useful to be able to do (like to host issuer info out of a bucket or something that doesn't have that directory structure)

aramase replied:
With this implementation, the behavior is same as using `issuer.url` for discovery; adding `/.well-known/openid-configuration` during runtime. The feature to [override discovered issuer URL in go-oidc](https://github.com/coreos/go-oidc/pull/315) was implemented similarly (_only overriding the host_), so I was leaning more towards what's generally used.

enj replied:
I don't have a strong opinion here, but the current implementation matches what is possible with the "insecure issuer URL" function of go-oidc v3.  If we care to be able to collapse onto that functionality any time in the future, then we need to handle appending the well known suffix.  But this round tripper is minimal enough that I am not worried about keeping it forever.

liggitt replied:
honestly, I find this roundtripper translation easier to reason about than creating an InsecureIssuerURLContext and wondering what is in the context and what it is used for

liggitt commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +362,28 @@ func New(opts Options) (*Authenticator, error) {
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, wellKnownEndpointPath) {
+		clone := req.Clone(req.Context())
```
do we need to override or clear req.Host?

do we need to check the whole URL of the intercepted request to make sure it matched `url`, not just a path suffix?

aramase replied:
> do we need to override or clear req.Host?

We're doing a deepcopy here and setting `clone.URL.Host` to the [new host](https://github.com/kubernetes/kubernetes/pull/123527/files#diff-de8a853b396792a685ba049763f700e10c7e03bc4482f6801e857e2f9a1b9f1eR379). That overrides the host?

> do we need to check the whole URL of the intercepted request to make sure it matched url, not just a path suffix?

The `go-oidc` library only makes the request to `/.well-known/openid-configuration` once as part of the new provider initialization and in case of the authenticator it's always with the configured issuer url. If we want to be exact here, we could check the url matches `opts.JWTAuthenticator.Issuer.URL + slashIfNeeded + wellKnownEndpointPath`. LMK if we should do that?

liggitt replied:
there's req.URL.Host and req.Host ... I don't have call paths clear enough in my head to know if the latter can be set as well

liggitt replied:
> If we want to be exact here, we could check the url matches opts.JWTAuthenticator.Issuer.URL + slashIfNeeded + wellKnownEndpointPath. LMK if we should do that?

I would check the whole thing

aramase replied:
I did a test and I do see `clone.Host` being overridden to `clone.URL.Host` after it's set.

liggitt replied:
the `req.Host` field takes precedence if set, I think:

https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/net/http/request.go;l=594-604

so maybe clear it from our clone?

aramase replied:
done!

liggitt commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -268,6 +273,25 @@ func New(opts Options) (*Authenticator, error) {
+		if baseTransport == nil {
+			baseTransport = http.DefaultTransport
+		}
+		clientWithDiscoveryURL.Transport = &discoveryURLRoundTripper{baseTransport, discoveryURL}
```
I think this needs to be told the URL to recognize (`opts.JWTAuthenticator.Issuer.URL + slashIfNeeded + wellKnownEndpointPath`)

aramase replied:
done!

aramase commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -175,9 +175,10 @@ type JWTAuthenticator struct {
+// Issuer provides the configuration for an external provider's specific settings.
 type Issuer struct {
 	URL                  string
+	DiscoveryURL         string
```
done!

aramase commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -220,6 +220,26 @@ type Issuer struct {
+	//     issuer: "https://oidc.example.com" (.url field)
+	// }
+	//
+	// Required to be unique.
```
I've added a comment about it requiring to be unique across JWT authenticators and required to be different from url.

liggitt commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -268,6 +273,28 @@ func New(opts Options) (*Authenticator, error) {
+		}
+		// This matches the url construction in oidc.NewProvider as of go-oidc v2.2.1.
+		// xref: https://github.com/coreos/go-oidc/blob/40cd342c4a2076195294612a834d11df23c1b25a/oidc.go#L114
+		urlToRewrite := strings.TrimSuffix(opts.JWTAuthenticator.Issuer.URL, "/") + wellKnownEndpointPath
```
can we construct the full target discovery URL at startup:

```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

aramase replied:
This results in
```bash
&url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"127.0.0.1:52645", Path:".well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
```

with the path missing the prefix slash.

liggitt replied:
I think that would actually work when it got stringified back out, but if you really want the leading slash in the internal representation, you can do this:

```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

aramase replied:
updated the PR with this

liggitt commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:
```diff
@@ -338,6 +365,33 @@ func New(opts Options) (*Authenticator, error) {
+	if req.Method == http.MethodGet && req.URL.String() == t.urlToRewrite {
+		clone := req.Clone(req.Context())
+		clone.Host = ""
+		clone.URL.RawPath = ""
```
if `t.discoveryURL` is fully constructed at startup, can we just do `clone.URL = t.discoveryURL` and override everything?

aramase replied:
This doesn't seem to work `clone.URL = t.discoveryURL`. Doing this results in `path` missing the leading slash.

```
before: &url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"auth.example.com", Path:"/.well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
after: &url.URL{Scheme:"https", Opaque:"", User:(*url.Userinfo)(nil), Host:"127.0.0.1:52528", Path:".well-known/openid-configuration", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
```

liggitt replied:
I think it would work with or without the leading slash (the stringification adds it)

aramase replied:
that's what I thought but the unit tests fail with `400 Bad Request` and I don't see the request come into the tests claim server with

```go
discoveryURL = discoveryURL.JoinPath(wellKnownEndpointPath)
clone.URL = t.discoveryURL
```

This works as expected and I've updated the PR with it
```go
discoveryURL.Path = path.Join(discoveryURL.Path, wellKnownEndpointPath)
```

aramase replied:
_Debugging `discoveryURL = discoveryURL.JoinPath(wellKnownEndpointPath)` further_

Dumping the request with
```go
	bytes, _ := httputil.DumpRequestOut(clone, true)
```

I see the path is without the leading slash
```bash
GET .well-known/openid-configuration HTTP/1.1
Host: 127.0.0.1:56692
User-Agent: Go-http-client/1.1
Accept-Encoding: gzip
```

liggitt replied:
that's super-weird... I'm surprised the go round tripper is happy writing an invalid request like that

aramase commented:
/retest

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go:
```diff
@@ -175,9 +175,40 @@ type JWTAuthenticator struct {
+	// certificateAuthority is used to verify the TLS connection and the hostname on the leaf certificate
+	// must be set to 'oidc.oidc-namespace'.
+	//
+	// curl https://oidc.oidc-namespace (.discoveryURL field)
```
If discoveryURL is `https://oidc.oidc-namespace`, then with the current documented behavior you would curl `https://oidc.oidc-namespace/.well-known/openid-configuration`, right?

I still think it would be clearer to make discoveryURL the full exact URL we fetch from. If we don't do that, I think we'll be adding a "useDiscoveryURLVerbatim bool" flag in a year to opt out of that behavior. People using discoveryURL are already advanced, giving them full exact control over the url to pull from without assuming the well-known suffix makes sense to me. We can document this very clearly with examples, like this:

```go
	// url points to the issuer URL in a format https://url or https://url/path.
  // ...
	// Discovery information is fetched from "{url}/.well-known/openid-configuration" unless overridden by discoveryURL.
  // ...
	URL string

  // discoveryURL, if specified, overrides the URL used to fetch discovery
  // information instead of using "{url}/.well-known/openid-configuration".
  // The exact value specified is used, so "/.well-known/openid-configuration"
  // must be included in discoveryURL if needed.
  // ...
  DiscoveryURL string
```

aramase replied:
Made `discoveryURL` the exact URL as per https://github.com/kubernetes/kubernetes/pull/123527#discussion_r1509619257. With this change I've removed the disallow trailing slash in the discovery url check as user has full control over it. LMK what you think!

liggitt commented:
impl looks good, last open question is whether to make discoveryURL completely authoritative or keep auto-appending `/.well-known/...`

aramase commented:
/retest

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -209,17 +209,45 @@ type JWTAuthenticator struct {
-	// Used to fetch discovery information unless overridden by discoveryURL.
-	// Required to be unique.
+	// Discovery information is fetched from "{url}/.well-known/openid-configuration" unless overridden by discoveryURL.
+	// Required to be unique across all JWT authenticators.
```
where are we validating URL is unique across all jwt authenticators? I don't see that in ValidateAuthenticationConfiguration

aramase replied:
Added a comment about there before: https://github.com/kubernetes/kubernetes/pull/123527#discussion_r1504997820

liggitt commented on staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1alpha1/types.go:
```diff
@@ -209,17 +209,45 @@ type JWTAuthenticator struct {
+	// }
+	//
+	// discoveryURL must be different from url.
+	// Required to be unique across all JWT authenticators.
```
where are we validating DiscoveryURL is unique across all jwt authenticators? I don't see that in ValidateAuthenticationConfiguration

aramase replied:
Previous comment: https://github.com/kubernetes/kubernetes/pull/123527#discussion_r1504996851

liggitt reviewed (commented):
impl and docs look good, looks like we need to validate the uniqueness as documented

aramase commented:
> impl and docs look good, looks like we need to validate the uniqueness as documented

Responded to the validation comments. We only support single JWT Authenticator as of now and I've a PR to support multiple (https://github.com/kubernetes/kubernetes/pull/123431). With that change, we will validate the unique issuer and discovery urls.

liggitt commented:
ah, ok

/lgtm
/approve

aramase commented:
`pull-kubernetes-unit` and `pull-kubernetes-verify-lint` were failing because https://github.com/kubernetes/kubernetes/pull/123225 merged and we changed `New` to return an interface.
```
staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go:319:16: a.idTokenVerifier undefined (type authenticator.Token has no field or method idTokenVerifier)
```

Rebased and updated `oidc_test.go` to include changes from https://github.com/kubernetes/kubernetes/pull/123225.

liggitt commented:
/lgtm
